frontend/src/wasm-communication/messages.ts

```diff
@@ -1188,13 +1190,46 @@ export type WidgetLayout = {
 	layout: LayoutGroup[];
 };
 
+export type WidgetDiffUpdate = {
+	layoutTarget: unknown;
+	diff: WidgetDiff[];
+};
+
+type WidgetDiff = { path: number[]; new: LayoutGroup[] | LayoutGroup | Widget | MenuBarEntry[] };
+
 export function defaultWidgetLayout(): WidgetLayout {
 	return {
 		layoutTarget: undefined,
 		layout: [],
 	};
 }
 
+// Updates a widget layout based on a list of updates, returning the new layout
+export function patchWidgetLayout(layout: WidgetLayout, updates: WidgetDiffUpdate): void {
+	layout.layoutTarget = updates.layoutTarget;
+
+	updates.diff.forEach((update) => {
+		// Find the object where the diff applies to
+		let targetLayout = layout.layout as LayoutGroup[] | LayoutGroup | Widget | MenuBarEntry[] | MenuBarEntry | undefined;
+		update.path.forEach((index) => {
```

Comment:
It looks like the goal of this loop is to find a single entry in the array, is that correct? Would it, thus, be desirable to break from the loop once it's found? You'd need to probably use a regular old loop in that case in order to break. Or do you intend to continue until the end of the loop every time?

Reply:
The path is a list of indexes i.e. the second widget in the third group in the first section would be a widgetPath of `[0, 2, 1]` We always want to go through each item of the path and sample the layoutTarget at that index.